### PR TITLE
Option to put TabContainer tabs at bottom

### PR DESCRIPTION
**Pull request by kitbdev**
part of: https://github.com/godotengine/godot-proposals/issues/1986
salvaged: https://github.com/godotengine/godot/pull/44420

Adds `tabs_at_bottom` to TabContainer.
Enabling will move the tab bar to the bottom.

image

### CODE REVIEW

**Review by YeldhamDev — CHANGES_REQUESTED** [review]
There should be a specific set of theme styles for the upside-down tabs.

**Comment by Maran23** [comment]
and especially for the future, this should be an enum. Right now it would only contain `TOP` and `BOTTOM`.  `TOP` would be the default and this way we do not break compatibility (also when adding the last two directions `LEFT` and `RIGHT` at one point)

**Comment by MewPurPur** [comment]
There is also a PR making it possible to add tabs to the left: #63873

**Comment by kitbdev** [comment]
Changed to use an enum `TabPosition`.
I used a theme variation for the bottom style `TabBarBottom` that is set when the position is set. TabContainer now sets its internal TabBar to use the same theme variation.
Editor theme variation (for testing):
image
Default theme variation:
image

File: scene/theme/default_theme.cpp [hunk 1]
```diff
@@ -925,6 +925,37 @@ void fill_default_theme(Ref<Theme> &theme, const Ref<Font> &default_font, const
...
+	theme->set_stylebox("tab_disabled", "TabBarBottom", style_tab_bottom_disabled);
+	theme->set_stylebox("tab_focus", "TabBarBottom", style_tab_bottom_focus);
+
+	theme->set_type_variation("TabBarBottom", "TabContainer");
+
+	theme->set_stylebox("tab_selected", "TabBarBottom", style_tab_bottom_selected);
+	theme->set_stylebox("tab_hovered", "TabBarBottom", style_tab_bottom_hovered);
+	theme->set_stylebox("tab_unselected", "TabBarBottom", style_tab_bottom_unselected);
+	theme->set_stylebox("tab_disabled", "TabBarBottom", style_tab_bottom_disabled);
+	theme->set_stylebox("tab_focus", "TabBarBottom", style_tab_bottom_focus);
```
Comment (YeldhamDev) [inline scene/theme/default_theme.cpp:957]: Those should be inside the `TabContainer` theme (with each style ending with "_bottom"), since it's there they are being used, and to allow editing them in the editor.

Comment (kitbdev) [reply scene/theme/default_theme.cpp:957]: I feel like type variations are a good fit for bottom tabs, since the tabs themselves shouldn't care about being on the bottom. It also simplifies the code and simplifies configuring styles for tabs. The TabBarBottom variation can also be configured separately in the theme editor.
If its about letting the user configure the variation, I could only set the variation if it hasn't been set, like:
```
if (tabs_position == POSITION_TOP) {
	if (get_theme_type_variation() == "TabBarBottom") {
		set_theme_type_variation("");
	}
} else {
	if (get_theme_type_variation() == "") {
		set_theme_type_variation("TabBarBottom");
	}
}
```
Then if the user wants to change the tab bar position, the theme variation is in their control.

**Comment by kitbdev** [comment]
fixed docs and don't override type variation if set

File: scene/theme/default_theme.cpp (hunk 1 above)
Comment (YeldhamDev) [reply scene/theme/default_theme.cpp:957]: The problem with this is that those styles will be hidden from the user, no way to know it via documentation, and no access via the inspector or the theme list. They would need to discover them from somewhere else somehow, and the only way to change them would be by creating a theme or doing it from code.

And besides all that, no other node does user-facing theming is this manner.

Comment (kitbdev) [reply scene/theme/default_theme.cpp:957]: We can improve the TabBar and TabContainer docs to mention when the theme variation should be used, like:
`tab_unselected`: "The style of the other, unselected tabs. If you need tabs at the [Constant POSITION_BOTTOM] and the [Constant POSITION_TOP], use `TabBarBottom` theme type variation to distinguish."

discoverability: 
when position changed to bottom in inspector: 
image
image
in the theme editor when adding a type:
image

As far as I know the only way to change the style without creating a theme or doing it from code is the style overrides, which works as expected.

Because of how variations work, when the regular TabContainer stylebox is changed and there is no TabBarBottom variation set, the style bottom tabs is set, so its not necessary to use the variation. 

Its not the same but, FileDialog and GraphEdit change some of their internal element's theme type variation, so if a user wants to change those parts they can use the type variations, and it isn't documented.

Comment (YeldhamDev) [reply scene/theme/default_theme.cpp:957]: > We can improve the TabBar and TabContainer docs to mention when the theme variation should be used

And what about being able to find those styles by simply searching for their names, like the theme items of other UI nodes?

> when position changed to bottom in inspector:

The way it's done here would actually create another problem on its own, what if the user is already using a theme override in their `TabContainer`? This would erase the value set by them without them knowing.

> Its not the same but, FileDialog and GraphEdit change some of their internal element's theme type variation

From what I saw, their variations are just empty styleboxes and small font tweaks. But even so, this shouldn't excuse going down that path in an even larger manner.

Comment (kitbdev) [reply scene/theme/default_theme.cpp:957]: > And what about being able to find those styles by simply searching for their names, like the theme items of other UI nodes?

The normal tab container tab styles (`tab_unselected`, etc.) are used to find the theme items when searching their names. If the user specifically wants to differentiate tabs at the bottom in a theme, the documentation for each of the theme items mentions the variation.
If the user is just trying to configure a tab container with tabs at the bottom, the normal theme items work fine. Changing between top and bottom tabs doesn't happen frequently enough that the theme items need to be separate.

Since tabs can't be at top and at bottom at the same time, having a theme variant is more intuitive than having extra theme items just for bottom tabs.

> what if the user is already using a theme override in their TabContainer?

The theme type variation does not get overridden if the user changed the value.

> From what I saw, their variations are just empty styleboxes and small font tweaks. But even so, this shouldn't excuse going down that path in an even larger manner.

I feel like this use case is very well suited to using theme type variations and should be allowed even if it sets a precedent.
Also, otherwise we would need to add a `use_bottom_style` flag to TabBar that only affects style and check it each time we draw a tab.

**Comment by kitbdev** [comment]
changed to use theme item for _bottom styles in tab container. I found a way to not have to modify TabBar at all using the theme overrides! That was a big concern I had with this method originally.

I am now looking into flipping the StyleBox automatically

**Comment by kitbdev** [comment]
Now automatically flips the tab's stylebox instead of using separate theme items.

**Review by YeldhamDev — APPROVED** [review]
LGTM

**Comment by kitbdev** [comment]
Rebased and fixed merge conflicts.
Removed unnecessary line: `tab_bar->set_theme_type_variation(get_theme_type_variation());`

**Review by AThousandShips — APPROVED** [review]

File: scene/gui/tab_bar.h [hunk 2]
```diff
@@ -210,6 +211,9 @@ class TabBar : public Control {
 	void set_clip_tabs(bool p_clip_tabs);
 	bool get_clip_tabs() const;
 
+	void set_tab_style_v_flip(bool p_tab_style_v_flip);
+	bool get_tab_style_v_flip() const;
```
Comment (KoBeWi) [inline scene/gui/tab_bar.h:215]: This seems unused. If something is not a property, it doesn't need a corresponding getter.

File: scene/gui/tab_container.cpp [hunk 3]
```diff
@@ -198,6 +220,8 @@ void TabContainer::_on_theme_changed() {
 		return;
 	}
 
+	tab_bar->set_tab_style_v_flip(tabs_position == POSITION_BOTTOM);
```
Comment (KoBeWi) [inline scene/gui/tab_container.cpp:223]: Why is this here if it doesn't depend on the theme?

**Comment by KoBeWi** [comment]
Probably not important, but there is a small difference between popup position in top and bottom tabs:
image
image
I think the bottom one looks better and the top one looks wrongly implemented (it uses icon height, not the actual header size).

File: scene/gui/tab_container.cpp [hunk 4]
```diff
@@ -165,21 +186,22 @@ void TabContainer::_notification(int p_what) {
...
-			int header_height = _get_top_margin();
+			int header_height = _get_tab_height();
+			int header_voffset = (tabs_position == POSITION_BOTTOM) * (size.height - header_height);
 
 			// Draw background for the tabbar.
-			theme_cache.tabbar_style->draw(canvas, Rect2(0, 0, size.width, header_height));
+			theme_cache.tabbar_style->draw(canvas, Rect2(0, header_voffset, size.width, header_height));
 			// Draw the background for the tab's content.
-			theme_cache.panel_style->draw(canvas, Rect2(0, header_height, size.width, size.height - header_height));
+			theme_cache.panel_style->draw(canvas, Rect2(0, (tabs_position == POSITION_TOP) * header_height, size.width, size.height - header_height));
```
Comment (KoBeWi) [inline scene/gui/tab_container.cpp:195]: Booleans in calculations are ok, but it's better if they are explicit IMO.
```suggestion
			theme_cache.panel_style->draw(canvas, Rect2(0, int(tabs_position == POSITION_TOP) * header_height, size.width, size.height - header_height));
```

File: scene/gui/tab_container.cpp [hunk 5]
```diff
@@ -890,6 +932,8 @@ void TabContainer::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("get_tab_control", "tab_idx"), &TabContainer::get_tab_control);
 	ClassDB::bind_method(D_METHOD("set_tab_alignment", "alignment"), &TabContainer::set_tab_alignment);
 	ClassDB::bind_method(D_METHOD("get_tab_alignment"), &TabContainer::get_tab_alignment);
+	ClassDB::bind_method(D_METHOD("set_tabs_position", "at_bottom"), &TabContainer::set_tabs_position);
```
Comment (KoBeWi) [inline scene/gui/tab_container.cpp:935]: `at_bottom` looks like a leftover.

File: scene/gui/tab_bar.cpp [hunk 6]
```diff
@@ -516,7 +516,13 @@ void TabBar::_draw_tab(Ref<StyleBox> &p_tab_style, Color &p_font_color, int p_in
 	bool rtl = is_layout_rtl();
 
 	Rect2 sb_rect = Rect2(p_x, 0, tabs[p_index].size_cache, get_size().height);
+	if (tab_style_v_flip) {
+		draw_set_transform(Point2(0.0, p_tab_style->get_draw_rect(sb_rect).size.y), 0.0, Size2((1.0), (-1.0)));
```
Comment (AThousandShips) [inline scene/gui/tab_bar.cpp:520]:
```suggestion
		draw_set_transform(Point2(0.0, p_tab_style->get_draw_rect(sb_rect).size.y), 0.0, Size2(1.0, -1.0));
```
Leftovers it appears, remove

File: scene/gui/tab_bar.cpp [hunk 7]
```diff
@@ -516,7 +516,13 @@ void TabBar::_draw_tab(Ref<StyleBox> &p_tab_style, Color &p_font_color, int p_in
...
+	}
 	p_tab_style->draw(ci, sb_rect);
+	if (tab_style_v_flip) {
+		draw_set_transform(Point2(), 0.0, Size2((1.0), (1.0)));
```
Comment (AThousandShips) [inline scene/gui/tab_bar.cpp:524]:
```suggestion
		draw_set_transform(Point2(), 0.0, Size2(1.0, 1.0));
```

**Comment by kitbdev** [comment]
Resolved issues.
I made the icon popup appear at the bottom or top of the menu icon. I don't just use the tab height in case it is large, the popup would look disconnected.
image
image

**Review by KoBeWi — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!
